File: app/models/payment.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, ForeignKey, Numeric, String, text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base

if TYPE_CHECKING:
    from app.models.order import Order


class Payment(Base):
    __tablename__ = "payments"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()")
    )
    order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("orders.id", ondelete="CASCADE"), nullable=False
    )
    method: Mapped[str] = mapped_column(String(30), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="pending")
    amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    external_reference: Mapped[str | None] = mapped_column(String(150))
    idempotency_key: Mapped[str | None] = mapped_column(String(80))
    paid_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=text("now()"))

    order: Mapped["Order"] = relationship(back_populates="payments")

File: php/src/Models/Payment.php
<?php

declare(strict_types=1);

namespace App\Models;

use App\Core\Money;

final class Payment
{
    public function __construct(
        public readonly string $id,
        public readonly string $orderId,
        public readonly string $method,
        public readonly string $status,
        public readonly int $amountCents,
        public readonly ?string $externalReference,
        public readonly ?string $paidAt,
    ) {
    }

    /** @param array<string, mixed> $row */
    public static function fromRow(array $row): self
    {
        return new self(
            id: $row['id'],
            orderId: $row['order_id'],
            method: $row['method'],
            status: $row['status'],
            amountCents: Money::fromDecimalString((string) $row['amount']),
            externalReference: $row['external_reference'],
            paidAt: $row['paid_at'],
        );
    }
}
